tests/Language/PureScript/Ide/UsageSpec.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+module Language.PureScript.Ide.UsageSpec where
+
+import           Protolude
+
+import qualified Data.Text as Text
+import           Language.PureScript.Ide.Command
+import           Language.PureScript.Ide.Types
+import qualified Language.PureScript.Ide.Test as Test
+import qualified Language.PureScript as P
+import           Test.Hspec
+import           Data.Text.Read (decimal)
+import System.FilePath
+
+load :: [Text] -> Command
+load = LoadSync . map Test.mn
+
+usage :: P.ModuleName -> Text -> IdeNamespace -> Command
+usage mn ident ns = FindUsages mn ident ns
+
+shouldBeUsage :: P.SourceSpan -> (FilePath, Text) -> Expectation
+shouldBeUsage usage' (fp, range) =
+  let
+    [ start, end] = Text.splitOn "-" range
+    unsafeReadInt = fst . either (panic "") identity . decimal
+    [ startLine, startColumn ] = map unsafeReadInt (Text.splitOn ":" start)
+    [ endLine, endColumn ] = map unsafeReadInt (Text.splitOn ":" end)
+  in
+    do
+      fp `shouldBe` P.spanName usage'
+
+      (startLine, startColumn)
+        `shouldBe`
+          ( P.sourcePosLine (P.spanStart usage')
+          , P.sourcePosColumn (P.spanStart usage'))
+      (endLine, endColumn)
+        `shouldBe`
+          ( P.sourcePosLine (P.spanEnd usage')
+          , P.sourcePosColumn (P.spanEnd usage'))
+
+spec :: Spec
+spec = describe "Finding Usages" $ do
+    it "finds a simple usage" $ do
+      ([_, Right (UsagesResult [usage1])], _) <- Test.inProject $
+        Test.runIde [ load ["FindUsage", "FindUsage.Definition", "FindUsage.Reexport"]
+                    , usage (Test.mn "FindUsage.Definition") "usageId" IdeNSValue
+                    ]
+      usage1 `shouldBeUsage` ("src" </> "FindUsage.purs", "12:11-12:18")
+    it "finds a constructor usage" $ do
+      ([_, Right (UsagesResult [usage1])], _) <- Test.inProject $
+        Test.runIde [ load ["FindUsage", "FindUsage.Definition", "FindUsage.Reexport"]
+                    , usage (Test.mn "FindUsage.Definition") "Used" IdeNSValue
+                    ]
+      usage1 `shouldBeUsage` ("src" </> "FindUsage.purs", "8:3-8:9")
+    it "finds a constructor alias usage" $ do
+      ([_, Right (UsagesResult [usage1])], _) <- Test.inProject $
+        Test.runIde [ load ["FindUsage", "FindUsage.Definition", "FindUsage.Reexport"]
+                    , usage (Test.mn "FindUsage.Definition") "$%" IdeNSValue
+                    ]
+      usage1 `shouldBeUsage` ("src" </> "FindUsage.purs", "9:3-9:9")
+    it "finds a reexported usage" $ do
+      ([_, Right (UsagesResult [usage1])], _) <- Test.inProject $
+        Test.runIde [ load ["FindUsage", "FindUsage.Definition", "FindUsage.Reexport"]
+                    , usage (Test.mn "FindUsage.Reexport") "toBeReexported" IdeNSValue
+                    ]
+        -- TODO(Christoph): Interesting parser bug here. It seems the position
```

Comment:
This is meant to be fixed in a separate PR? Should there be an issue for it?

Reply:
I'll probably want to confirm this first with a few more test cases. It's definitely not something I'd want to fix in this PR.

Reply:
I think I've seen this bug too - usually in the atom plugin, if the last line has a warning/error on it, the span ends early.

Reply:
This is because when we draw a token, the end span is determined by the start of the next token. Since the last token has no next token, the span is incorrect. I looked into this a while ago when looking into fixing positions for psa (since the span also includes whitespace and comments which are consumed as part of the lexeme grammar) and it is non trivial.

Reply:
I ended up just using a bunch of heuristics in psa to clean it up.

Reply:
I tried to fix this in the parser and failed utterly.

Reply:
Yes it basically amounts to completely rethinking how the parser works, unfortunately, or sacrificing substantial performance.

Reply:
Is there anywhere these struggles could be documented? It would probably be good to have this knowledge easier to find for the next person who tries to deal with this, no?

Reply:
FWIW there’s this https://github.com/purescript/purescript/issues/1702